### PR TITLE
feat(ingest/bigquery): Support for View Labels

### PR DESCRIPTION
**Pull request by ethan-cartwright**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

**Comment by ethan-cartwright** [comment]
todo: add option_name = "labels" to get views in the queries.py eg.

```
SELECT
  t.table_catalog as table_catalog,
  t.table_schema as table_schema,
  t.table_name as table_name,
  t.table_type as table_type,
  t.creation_time as created,
  ts.last_modified_time as last_altered,
  tos.OPTION_VALUE as comment,
  t.is_insertable_into,
  t.ddl as view_definition,
  ts.row_count,
  ts.size_bytes
FROM
  `acryl-staging.ethan_test_dataset.INFORMATION_SCHEMA.TABLES` t
  JOIN `acryl-staging.ethan_test_dataset.__TABLES__` as ts ON ts.table_id = t.TABLE_NAME
  LEFT JOIN `acryl-staging.ethan_test_dataset.INFORMATION_SCHEMA.TABLE_OPTIONS` as tos ON t.table_schema = tos.table_schema
  AND t.TABLE_NAME = tos.TABLE_NAME
  AND tos.OPTION_NAME = "labels"
WHERE
  table_type in ('VIEW', 'MATERIALIZED_VIEW')
  AND t.table_name = 'account_types'
ORDER BY
  table_schema ASC,
  table_name ASC;
```

**Comment by ethan-cartwright** [comment]
Screenshot 2024-06-13 at 7 55 54 PM

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py [hunk 1]
```diff
@@ -1048,11 +1048,22 @@ def gen_view_dataset_workunits(
...
+        tags_to_add = None
+        if table.labels and self.config.capture_view_label_as_tag:
+            tags_to_add = []
+            tags_to_add.extend(
+                [
+                    make_tag_urn(f"""{k}:{v}""")
+                    for k, v in table.labels.items()
+                    if is_tag_allowed(self.config.capture_view_label_as_tag, k)
+                ]
+            )
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py:1060]:
```suggestion
            tags_to_add = [
                make_tag_urn(f"{k}:{v}")
                for k, v in table.labels.items()
                if is_tag_allowed(self.config.capture_view_label_as_tag, k)
            ]
```

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_helper.py [hunk 2]
```diff
@@ -17,3 +18,13 @@ def unquote_and_decode_unicode_escape_seq(
...
+
+
+def parse_labels(labels_str: str) -> Dict[str, str]:
+    pattern = r'STRUCT\("([^"]+)", "([^"]+)"\)'
+
+    # Find all matches in the labels string
+    matches = re.findall(pattern, labels_str)
+
+    labels_dict = {key: value for key, value in matches}
+    return labels_dict
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_helper.py:30]:
```suggestion
    # Map ?? to ??
    return dict(re.findall(pattern, labels_str))
```
I don't actually know what the keys and values of the returned dict represent. Let's document that somewhere

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py [hunk 3]
```diff
@@ -51,9 +52,9 @@ def from_time_partitioning(
         cls, time_partitioning: TimePartitioning
     ) -> "PartitionInfo":
         return cls(
-            field=time_partitioning.field
-            if time_partitioning.field
-            else "_PARTITIONTIME",
+            field=(
+                time_partitioning.field if time_partitioning.field else "_PARTITIONTIME"
+            ),
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py:57]:
```suggestion
            field=time_partitioning.field or "_PARTITIONTIME",
```

**Review by asikowitz — APPROVED** [review]
Just some style comments. Otherwise LGTM although would like a second look on the SQL query

**Comment by ethan-cartwright** [comment]
> Just some style comments. Otherwise LGTM although would like a second look on the SQL query

@hsheth2 can you take a look?

**Comment by ethan-cartwright** [comment]
> > Just some style comments. Otherwise LGTM although would like a second look on the SQL query
> 
> @hsheth2 can you take a look?

Harshal took a look and gave the go ahead. https://acryldata.slack.com/archives/C02ENS87APK/p1718410928205629?thread_ts=1718392351.832049&cid=C02ENS87APK

**Comment by anshbansal** [comment]
Airflow problem is separate

**Pull request merged**